File: preparation/src/product/domain/events/ProductCreated.cs
/* 
 * 
 * Libraries
 *
*/

using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.product.domain {

    public class ProductCreated : DomainEvent {

        /* 
         * 
         * Attributes
         *
        */

        private ProductId   id;
        private ProductName name;

        /* 
         * 
         * Methods
         *
        */

        public ProductCreated() {
        }

        public ProductCreated( ProductId id, ProductName name ) {
            this.id   = id;
            this.name = name;
        }

        public ProductCreated(
            ProductId   id,
            ProductName name,
            string      eventId,
            int         timestamp
        ) : base( eventId, timestamp ) {
            this.id   = id;
            this.name = name;
        }

        public ProductId getId() {
            return id;
        }

        public ProductName getName() {
            return name;
        }

        public override DomainEvent fromPrimitives(
            string                     eventId, 
            int                        timestamp, 
            Dictionary<string, object> data
        ) {
            // Variables
            ProductCreated domainEvent;
            // Code
            domainEvent = new ProductCreated(
                new ProductId( data["id"].ToString() ),
                new ProductName( data["name"].ToString() ),
                eventId,
                timestamp
            );
            return domainEvent;
        }

        public override string getEventName() {
            return "product.created";
        }

        public override Dictionary<string, object> toPrimitives() {
            // Variables
            Dictionary<string, object> data;
            // Code
            data = new Dictionary<string, object>();
            data.Add( "id", id.getValue() );
            data.Add( "name", "" + name.getValue() );
            return data;
        }

    }

}

File: backoffice/src/product/domain/events/ProductCreated.py
"""
 *
 * Libraries 
 *
"""

from ..valueObjects.ProductName        import ProductName
from ..valueObjects.ProductPrice       import ProductPrice
from ..valueObjects.ProductDescription import ProductDescription
from ..valueObjects.ProductId          import ProductId
from src.restaurant.domain             import RestaurantId
from src.shared.domain                 import DomainEvent

"""
 *
 * Class
 *
"""

class ProductCreated( DomainEvent ):

    """
     *
     * Attributes 
     *
    """

    __id           : ProductId
    __name         : ProductName
    __price        : ProductPrice
    __description  : ProductDescription
    __restaurantId : RestaurantId

    """
     *
     * Methods 
     *
    """

    def __init__( 
        self, 
        id           : ProductId          = None, 
        name         : ProductName        = None,
        price        : ProductPrice       = None,
        description  : ProductDescription = None,
        restaurantId : RestaurantId       = None,
        eventId      : str                = None,
        timestamp    : int                = None
    ) -> None:
        super().__init__( eventId, timestamp )
        self.__id           = id
        self.__name         = name
        self.__price        = price
        self.__description  = description
        self.__restaurantId = restaurantId
    
    def getId( self ) -> ProductId:
        return self.__id
    
    def getName( self ) -> ProductName:
        return self.__name
    
    def getPrice( self ) -> ProductPrice:
        return self.__price

    def getDescription( self ) -> ProductDescription:
        return self.__description
    
    def getRestaurantId( self ) -> RestaurantId:
        return self.__restaurantId
    
    def getEventName( self ) -> str:
        return 'product.created'
    
    def toPrimitives( self ) -> dict[str, str | int]:
        # Variables
        data : dict[str, str | int]
        # Code
        data = {
            'id'           : self.__id.getValue(),
            'name'         : self.__name.getValue(),
            'price'        : self.__price.getValue(),
            'description'  : self.__description.getValue(),
            'restaurantId' : self.__restaurantId.getValue()
        }
        return data

    def fromPrimitives( self, 
        eventId   : str, 
        timestamp : int, 
        data      : dict[str, str | int]
    ) -> object:
        # Variables
        domainEvent : object
        # Code
        domainEvent = ProductCreated(
            ProductId( data['id'] ),
            ProductName( data['name'] ),
            ProductPrice( data['price'] ),
            ProductDescription( data['description'] ),
            RestaurantId( data['restaurantId'] ),
            eventId,
            timestamp
        )
        return domainEvent

